k2/csrc/ragged_ops.cu

```diff
@@ -793,4 +794,72 @@ RaggedShape TrivialShape(ContextPtr &c, int32_t num_elems) {
   return RaggedShape2(&row_splits, &row_ids, num_elems);
 }
 
+// This function  implements the transpose algorithm for sparse matrices from
+// the following paper:
+//
+// Gustavson, Fred G.
+// "Two fast algorithms for sparse matrices: Multiplication and permuted
+// transposition." ACM Transactions on Mathematical Software (TOMS) 4.3 (1978):
+// 250-269.
+//
+Array1<int32_t> GetTransposeReordering(Ragged<int32_t> &src, int32_t num_cols) {
+  if (src.Context()->GetDeviceType() != kCpu) {
+    K2_LOG(FATAL) << "GPU support will be added later";
+  }
+
+  ContextPtr &context = src.Context();
+  int32_t num_axes = src.NumAxes();
+  Array1<int32_t> &row_splits = src.RowSplits(num_axes - 1);
+
+  int32_t num_rows = row_splits.Dim() - 1;
+  const int32_t *row_ptr = row_splits.Data();
+
+  // add an extra element for ExclusiveSum
+  Array1<int32_t> column_counts(context, num_cols + 1, 0);
+  int32_t *col_ptr = column_counts.Data();
+
+  const Array1<int32_t> &values = src.values;
+  const int32_t *values_ptr = values.Data();
+
+  auto lambda_column_counts = [=] __host__ __device__(int32_t i) -> void {
+    const int32_t *begin = values_ptr + row_ptr[i];
+    const int32_t *end = values_ptr + row_ptr[i + 1];
+    for (; begin != end; ++begin) {
```

Comment:
As `end-begin` may be large, I'm afraid the code here will be relatively too slow. At least I think we can launch a kernel and get the index (`*begin` here) in each thread, instead of in a loop.

Reply:
I mean, at least we can do (in a kernel) 
```
int32_t row = threadIdx / num_threads_per_row;
int32_t thread_in_this_row = threadIdx % num_threads_per_row;
```

Reply:
yeah, the way you would do this is to have the kernel parameter i be the idx01, then use row_ids to get the idx0 then row_splits to get the idx0x and then the idx1.  Of course incrementing the column pointer would need to be protected: could use atomicAdd but would need to write host version as well to avoid compilation error.  (Or we can make a version of Eval() that's GPU only.
However there are certain inputs for which that algorithm would be very slow, because the atomicAdd makes it sequential.  That's why it's a bit safer to do it based on sorting.